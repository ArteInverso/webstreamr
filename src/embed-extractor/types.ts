import { StreamWithMeta } from '../types';

export interface EmbedExtractor {
  readonly id: string;

  readonly label: string;

  readonly extract: (embedUrl: string, language: string) => Promise<StreamWithMeta>;
}
