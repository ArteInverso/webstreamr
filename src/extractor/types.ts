import { Context, Meta, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (ctx: Context, url: URL) => boolean;

  readonly extract: (ctx: Context, url: URL, meta: Meta) => Promise<UrlResult[]>;
}
