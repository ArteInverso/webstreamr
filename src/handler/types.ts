import { Context, UrlResult } from '../types';

export interface Handler {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: string[];

  readonly languages: string[];

  readonly handle: (ctx: Context, type: string, id: string) => Promise<(UrlResult | undefined)[]>;
}
