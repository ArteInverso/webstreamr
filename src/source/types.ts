import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Id } from '../utils';

export interface SourceResult {
  countryCode: CountryCode;
  title?: string;
  url: URL;
}

export interface Source {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: ContentType[];

  readonly countryCodes: CountryCode[];

  handle(ctx: Context, type: ContentType, id: Id): Promise<(SourceResult[])>;
}
