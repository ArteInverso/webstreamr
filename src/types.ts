import { Manifest, ManifestConfig } from 'stremio-addon-sdk';

export interface Context {
  id: string;
  ip: string;
  config: Config;
}

export type ManifestWithConfig = Manifest & { config: ManifestConfig[] };

export type Config = Record<string, string>;

export interface Meta {
  bytes?: number;
  countryCode: string;
  height?: number;
}

export interface UrlResult {
  url: URL;
  isExternal?: boolean;
  label: string;
  sourceId: string;
  meta: Meta;
  requestHeaders?: Record<string, string>;
}
