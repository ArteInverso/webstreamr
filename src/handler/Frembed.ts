import { ContentType } from 'stremio-addon-sdk';
import { Handler, HandleResult } from './types';
import { Fetcher, getTmdbId, Id } from '../utils';
import { Context, CountryCode } from '../types';

export class Frembed implements Handler {
  readonly id = 'frembed';

  readonly label = 'Frembed';

  readonly contentTypes: ContentType[] = ['series'];

  readonly countryCodes: CountryCode[] = [CountryCode.fr];

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly handle = async (ctx: Context, _type: string, id: Id): Promise<HandleResult[]> => {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);

    const apiUrl = new URL(`https://frembed.space/api/series?id=${tmdbId.id}&sa=${tmdbId.season}&epi=${tmdbId.episode}&idType=tmdb`);

    const json = JSON.parse(await this.fetcher.text(ctx, apiUrl));

    const urls: URL[] = [];
    for (const key in json) {
      if (key.startsWith('link') && json[key] && !json[key].includes(',https')) {
        try {
          urls.push(new URL(json[key].trim()));
        } catch {
          // Skip invalid URL
        }
      }
    }

    return urls.map(url => ({ countryCode: CountryCode.fr, referer: apiUrl, title: `${json['title']} ${tmdbId.season}x${tmdbId.episode}`, url }));
  };
}
