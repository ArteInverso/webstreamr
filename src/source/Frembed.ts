import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Fetcher, getTmdbId, Id } from '../utils';
import { Source, SourceResult } from './types';

export class Frembed implements Source {
  public readonly id = 'frembed';

  public readonly label = 'Frembed';

  public readonly contentTypes: ContentType[] = ['series'];

  public readonly countryCodes: CountryCode[] = [CountryCode.fr];

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  public async handle(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);

    const apiUrl = new URL(`https://frembed.wiki/api/series?id=${tmdbId.id}&sa=${tmdbId.season}&epi=${tmdbId.episode}&idType=tmdb`);

    const json = JSON.parse(await this.fetcher.text(ctx, apiUrl));

    const urls: URL[] = [];
    for (const key in json) {
      if (key.startsWith('link') && json[key] && !json[key].includes(',https')) {
        try {
          urls.push(new URL(json[key].trim()));
        } catch {
          // Skip invalid URL
        }
      }
    }

    return urls.map(url => ({ countryCode: CountryCode.fr, title: `${json['title']} ${tmdbId.season}x${tmdbId.episode}`, url }));
  };
}
