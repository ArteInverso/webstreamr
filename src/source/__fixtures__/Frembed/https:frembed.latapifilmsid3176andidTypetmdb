{"tmdb":"3176","title":"Battle Royale","link1":"https://d-s.io/e/gtenfs0rffzq","link2":"https://netu.fanstream.live/e/p0QUHavRbQD0","link3":"https://jilliandescribecompany.com/e/lbeqm6ofmauq","link4":"https://video.streamtales.cc/player/frvod.php?url=https://streamtales.cc:8443/videos/3176.mp4","link5":null,"link6":"","link7":"https://uqload.cx/embed-pnxmokgrigu9.html","link1vostfr":null,"link2vostfr":null,"link3vostfr":null,"link4vostfr":null,"link5vostfr":null,"link6vostfr":null,"link7vostfr":null,"link1vo":null,"link2vo":null,"link3vo":null,"link4vo":null,"link5vo":null,"link6vo":null,"link7vo":null,"imdb":"tt0266308","year":"2000","version":"TrueFrench","quality":"HD","qualityVostfr":null,"qualityVo":null,"poster":"/8PzexNbQ9eAKxe0p2U4XP3DVJCp.jpg","backdrop":"/pmRtI4QvcBvXjw9mLOEAsV9Bq7i.jpg,/amBvmIshdsSkOtvVIgxl7YSQ9Dg.jpg,/n4rNQUV9S9DMgCkQVt7R7VLYxNr.jpg,/55qVIFhykb6M6CGWRKQd9BsLI5D.jpg,/e9vJTJHkCOyBitE67n8zlp5Qm8q.jpg,/tSbZky1KLzqOSGPzYtqnxTQzNFy.jpg,/cQR3VHRMUL8xGkWK5Up8rI7DHQD.jpg,/4oIoe9tcSRzGSImp9Jqp1SKFwes.jpg,/zajlFkhKugj9tqH57KHXpmTXIku.jpg,/hpP0ELkViFlU8PVWbEBvVx12YtJ.jpg","checked":null,"dead_link":0,"views":"4","date_creation":"2023-08-17T04:06:45.000Z","update_date":"2024-09-19T01:28:34.000Z","original_title":"バトル・ロワイアル","title_vf":"Battle Royale","tatavid":null,"dcsutom":96950}