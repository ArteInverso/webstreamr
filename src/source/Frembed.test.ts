import winston from 'winston';
import { Frembed } from './Frembed';
import { Fetcher, ImdbId, TmdbId } from '../utils';
import { Context } from '../types';

jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const fetcher = new Fetcher(logger);
const ctx: Context = { id: 'id', ip: '127.0.0.1', config: { fr: 'on' } };

describe('Frembed', () => {
  let handler: Frembed;

  beforeEach(() => {
    handler = new Frembed(fetcher);
  });

  test('handle imdb black mirror s4e2', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt2085059', 4, 2));
    expect(streams).toMatchSnapshot();
  });

  test('handle tmdb black mirror s4e2', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(42009, 4, 2));
    expect(streams).toMatchSnapshot();
  });
});
