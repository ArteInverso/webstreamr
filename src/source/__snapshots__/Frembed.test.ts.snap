// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Frembed handle imdb black mirror s4e2 1`] = `
[
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://do7go.com/e/dfx8me4un4ul",
  },
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://netu.fanstream.us/e/0DFgfkcXOsDP",
  },
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://johnalwayssame.com/e/cqy9oue7sv0g",
  },
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://ds2play.com/e/fzfvfq3ngig0",
  },
]
`;

exports[`Frembed handle tmdb black mirror s4e2 1`] = `
[
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://do7go.com/e/dfx8me4un4ul",
  },
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://netu.fanstream.us/e/0DFgfkcXOsDP",
  },
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://johnalwayssame.com/e/cqy9oue7sv0g",
  },
  {
    "countryCode": "fr",
    "referer": "https://frembed.space/api/series?id=42009&sa=4&epi=2&idType=tmdb",
    "title": "Black Mirror 4x2",
    "url": "https://ds2play.com/e/fzfvfq3ngig0",
  },
]
`;
